### ISSUE TITLE
VideoStreamPlayer not drawing anything when stopped, including not showing in editor

### ISSUE THREAD

**Issue by lostminds** [godotengine/godot#92050]
### Tested versions

v4.3.dev6.official [89850d553]

### System information

macOS 14.5

### Issue description

It appears the VideoStreamPlayer doesn't render any content when stopped, so it's invisible until playing. What I'd expect is that when a video stream is loaded, it would show the first frame of this video. In the editor this is a bit confusing as it appears like the video isn't loaded and makes layouting more difficult, and in the running project it means you have to do a couple of extra steps to get the video to show in a paused state before you start playing it. 

![Screenshot 2024-05-17 at 15 42 02](https://github.com/godotengine/godot/assets/17763524/a53e0341-04c2-4023-9a1a-6306f2fdbf83)

I can tell that the video is loaded since the control locks its' minimum size to the video size. If you run the project it's the same, the video player is blank until you call `play()` on the player but then the video displays as expected. If you then call `stop()` it becomes blank again.

![Screenshot 2024-05-17 at 15 42 14](https://github.com/godotengine/godot/assets/17763524/cf39bd3a-653f-4ce3-a942-fa96ba24c487)

Once you've played at least one frame you can use `paused = true` to have the video display and not be playing. However, it seems it needs to play at least a little to load a first frame, so setting the player to `autoplay = true` and `paused = true` initially doesn't work. And in the editor there's no way to start playing as far as I can tell.

### Steps to reproduce

- Add a VideoStreamPlayer to a UI scene
- Set the stream to a supported video file (needs to be a Ogg Theora .ogv file), for example this file: [test_video.ogv](https://www.lostminds.com/godot/test_video.ogv)
- Observe the VideoStreamPlayer size change, but nothing showing

### Minimal reproduction project (MRP)

[video_player_test.zip](https://github.com/godotengine/godot/files/15350385/video_player_test.zip)

**Comment by AThousandShips** [comment]
This might be a limitation of the implementation, and might depend on processing and resources, unsure if this is by design, but the video steam feature is somewhat lacking and is being considered for moving out of the engine entirely, so it's unclear what exactly might be the cause of this

**Comment by lostminds** [comment]
> the video steam feature is somewhat lacking and is being considered for moving out of the engine entirely

Yes, I noticed it was a little particular with the format/encoding requirement and I saw that support for another format was dropped in 4.0. I think that some way of showing prerendered cutscenes or intros may still be a very useful feature in many projects, as well as perhaps some more light-weight solution for showing smaller animations/videos without audio for in-game tutorials, demos of skills/abilities etc. So if the video player is moved out of the core, perhaps it could be a reason to re-consider support for loading simple animation formats like GIF and APNG (like in https://github.com/godotengine/godot-proposals/issues/1433)

**Comment by AThousandShips** [comment]
The problem is that playing videos in an open source manner is surprisingly difficult to maintain, it's a tangle of libraries and compatibility issues and even maintaining one format has proven difficult, see [here](https://github.com/godotengine/godot-proposals/discussions/9669#discussioncomment-9425717) for example

It'd depend on if this is simply a mistake in the code for the theora code, missing a frozen frame, or if it's a tricky issue with how it decodes frames, I unfortunately can't answer that since I don't know how that particular code or library works

**Comment by lostminds** [comment]
> The problem is that playing videos in an open source manner is surprisingly difficult to maintain

I'm sure this is the case, and as mentioned in https://github.com/godotengine/godot-proposals/issues/7062#issuecomment-1586328263 I think a better way forward would then be to fall back to system/OS specific video streaming solutions so it's not needed to be handled by Godot and can also benefit from .mp4/H.264 support and better hardware acceleration.

However, in this particular issue I don't think the problem is with the formats or encoding. When playing the video it displays just fine as far as I can tell. It's just that the video isn't loaded or displayed before you play (or after you've stopped it), and in the editor you can't play it. Either this is accidental or it's by design to save performance and resources by not preparing the video for display before it's started, but either way as I've described above I think it leads to some undesirable outcomes, especially when working in the editor.

A work-around of sorts to the editor display issue without needing to load or generate an initial still frame to display in the video could be to just add play and stop buttons to allow playing VideoStreamPlayers in the editor. I think a PR a while back added the functionality to add action buttons in the inspector, which could be used for this?

**Comment by berarma** [comment]
This is done by design. I'm working on [fixing Theora video playing in Godot](#101958). I could work on this issue but I'd need to know what the correct behavior would be.

VideoStreamTheora intentionally puts an empty frame when the video isn't playing. It even puts the empty frame when looping the video. I think the latter is definitely wrong and could be fixed by simply removing this empty frame between plays.

But what should be shown in the editor or when the video is not playing?

**Comment by lostminds** [comment]
> But what should be shown in the editor or when the video is not playing?

I'd say that the VideoStreamPlayer should ideally show the frame at the current `stream_position` time, and update with changes to this property, even if it's not playing. This would then be the same in the editor as well, where the player is stopped.

If this is difficult to get working, just showing the first frame would I think be much better than the current blank to get a preview of the video content working with a VideoSteamPlayer in a UI layout.

**Comment by lostminds** [comment]
> In the playing/paused state it shows a video frame but in the stopped state it becomes invisible. Should we keep it this way? Because keeping it visible when it's stopped might break some uses where we want it to disappear when it isn't playing.

As implied above I thought this was a bug since I expected it to show at least the first frame or some poster frame when stopped, both in the editor and in game. But for compatibility perhaps a new `show_when_stopped` property could then be added controlling if some frame (preferably the current) should be shown when the video player is stopped. If the property defaults to `false` it shouldn't break any compatibility. This would also have the benefit of making the appearance in editor and in game the same.

**Comment by berarma** [comment]
There have been many changes in the VideoStreamPlayer and this issue is partially outdated. With Godot 4.5, stopping a video won't remove the current frame, similar to pause. But the VideoStreamPlayer is still invisible until the video starts playing.

It's transparent before playback starts, but after that, even when stopping the video, it will be showing the last frame. Maybe this is a bit incongruous. Anyway, I don't feel like changing it now because it might break some uses.

Fortunately, there's an easy workaround that will work with the next dev build. Use this code in your VideoStreamPlayer nodes:

```
@tool
extends VideoStreamPlayer

func _ready():
	if Engine.is_editor_hint():
	        play()

func _process(_delta: float):
	if Engine.is_editor_hint():
		stop()
```
